deployment/lib/lambdas/cf-url-rewriter.ts

```diff
@@ -0,0 +1,70 @@
+import { CloudFrontRequestEvent, CloudFrontRequest } from 'aws-lambda';
+
+const https = require('https');
+
+export async function handler(event: CloudFrontRequestEvent, context, callback): Promise<CloudFrontRequest> {
+  const request = event.Records[0].cf.request;
+
+  console.log('request', JSON.stringify(request));
+
+  if (request.uri.includes("latest")) {
```

Comment:
The regex/string has to match twice, which is wasteful. [JavaScript replace](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) takes a replacer function that is called on match, so you can do a single `request.uri.replace(/\/latest\/.*/, function(match, ...) { ... }`.

Reply:
Experiment a bit and try to create a function `getLatestNumber` and then put it in the replace call `request.uri.replace(/\/latest\/.*/, getLatestNumber);`. However, it is hard to decide whether a 302 redirect shall happen since this matching information is encapsulated inside `getLatestNumber`. Furthermore, some error handling becomes complicated since it needs a way to surface up such info from `getLatestNumber`.

Below is the draft code when I try to introduce `getLatestNumber` as a reference for my comment above.

```
const https = require('https');


exports.handler = async (event, context, callback) => {

  const request = event.Records[0].cf.request;
  
  // console.log('replaced', request.uri.replace(/\/latest\/.*/, test));
  request.uri = request.uri.replace(/^\/ci\/...\//, '\/');

  request.uri.replace(/\/latest\/.*/, getLatestNumber);

  callback(null, request);
}

async function getLatestNumber(request, callback){
    console.log(request);

    const indexUri = request.uri.replace(/\/latest\/.*/, '/index.json');
    
    console.log(indexUri);

    try {
      const data = await httpGet('https://' + request.headers.host[0].value + indexUri);

      if (data && data.latest) {
        return data.latest;
        // callback(null, redirectResponse(request, data.latest));
      } else {
        callback(null, errorResponse());
      }
    } catch (e) {
      console.error(e);
      callback(null, errorResponse());
    }
  
}

function test(match){
  console.log('run it', match);
  
  return 133;
}

function redirectResponse(request, latestNumber) {
  return {
    status: '302',
    statusDescription: 'Moved temporarily',
    headers: {
      'location': [{
        key: 'Location',
        value: request.uri.replace('latest', latestNumber),
      }],
      'cache-control': [{
        key: 'Cache-Control',
        value: "max-age=3600"
      }],
    },
  };

}

function errorResponse() {
  return {
    body: 'The page is not found!',
    status: '404',
    statusDescription: 'Not found',
  };
}

async function httpGet(url) {
  return new Promise((resolve, reject) => {

    https.get(url, (res) => {
      let body = "";

      res.on("data", (chunk) => {
        body += chunk;
      });

      res.on("end", () => {
        try {
          let json = JSON.parse(body);

          console.log("json ", json);
          resolve(json);

          // do something with JSON
        } catch (error) {
          console.error(error.message);
          reject(error);
        };
      });

    }).on("error", (error) => {
      console.error(error.message);
      reject(error);

    });
  });
}
```